File: algorithms/trees/trie/typescript/trieInsertSearch.ts
class TrieNode {
    children: Map<string, TrieNode> = new Map();
    isEnd: boolean = false;
}

function insert(root: TrieNode, key: number): void {
    let node = root;
    for (const ch of String(key)) {
        if (!node.children.has(ch)) {
            node.children.set(ch, new TrieNode());
        }
        node = node.children.get(ch)!;
    }
    node.isEnd = true;
}

function search(root: TrieNode, key: number): boolean {
    let node = root;
    for (const ch of String(key)) {
        if (!node.children.has(ch)) {
            return false;
        }
        node = node.children.get(ch)!;
    }
    return node.isEnd;
}

export function trieInsertSearch(arr: number[]): number {
    const n = arr.length;
    const mid = Math.floor(n / 2);
    const root = new TrieNode();

    for (let i = 0; i < mid; i++) {
        insert(root, arr[i]);
    }

    let count = 0;
    for (let i = mid; i < n; i++) {
        if (search(root, arr[i])) {
            count++;
        }
    }

    return count;
}

File: algorithms/trees/trie/rust/trie_insert_search.rs
use std::collections::HashMap;

struct TrieNode {
    children: HashMap<u8, TrieNode>,
    is_end: bool,
}

impl TrieNode {
    fn new() -> Self {
        TrieNode {
            children: HashMap::new(),
            is_end: false,
        }
    }
}

fn insert(root: &mut TrieNode, key: i32) {
    let s = key.to_string();
    let mut node = root;
    for &ch in s.as_bytes() {
        node = node.children.entry(ch).or_insert_with(TrieNode::new);
    }
    node.is_end = true;
}

fn search(root: &TrieNode, key: i32) -> bool {
    let s = key.to_string();
    let mut node = root;
    for &ch in s.as_bytes() {
        match node.children.get(&ch) {
            Some(child) => node = child,
            None => return false,
        }
    }
    node.is_end
}

pub fn trie_insert_search(arr: &[i32]) -> i32 {
    let n = arr.len();
    let mid = n / 2;
    let mut root = TrieNode::new();

    for i in 0..mid {
        insert(&mut root, arr[i]);
    }

    let mut count = 0;
    for i in mid..n {
        if search(&root, arr[i]) {
            count += 1;
        }
    }

    count
}
